lib/evaluate/integer.h

```diff
@@ -0,0 +1,853 @@
+        }
+      }
+    }
+    for (; j < parts; ++j) {
+      result.LEPart(j) = 0;
+    }
+    return result;
+  }
+
+  static constexpr ValueWithOverflow ReadUnsigned(
+      const char *&pp, std::uint64_t base = 10) {
+    Integer result;
+    bool overflow{false};
+    const char *p{pp};
+    while (*p == ' ' || *p == '\t') {
+      ++p;
+    }
+    if (*p == '+') {
+      ++p;
+    }
+    Integer radix{base};
+    for (; std::uint64_t digit = *p; ++p) {
+      if (digit >= '0' && digit < '0' + base) {
+        digit -= '0';
+      } else if (base > 10 && digit >= 'A' && digit < 'A' + base - 10) {
+        digit -= 'A' - 10;
+      } else if (base > 10 && digit >= 'a' && digit < 'a' + base - 10) {
+        digit -= 'a' - 10;
+      } else {
+        break;
+      }
+      Product shifted{result.MultiplyUnsigned(radix)};
+      overflow |= !shifted.upper.IsZero();
+      ValueWithCarry next{shifted.lower.AddUnsigned(Integer{digit})};
+      overflow |= next.carry;
+      result = next.value;
+    }
+    pp = p;
+    return {result, overflow};
+  }
+
+  template<typename FROM>
+  static constexpr ValueWithOverflow ConvertUnsigned(const FROM &that) {
+    std::uint64_t field{that.ToUInt64()};
+    ValueWithOverflow result{field, false};
+    if constexpr (bits < 64) {
+      if ((field >> bits) != 0) {
+        result.overflow = true;
+        return result;
+      }
+    }
+    for (int j{64}; j < that.bits; j += 64) {
+      field = that.SHIFTR(j).ToUInt64();
+      if (bits <= j) {
+        if (field != 0) {
+          result.overflow = true;
+          break;
+        }
+      } else {
+        result.value = result.value.IOR(Integer{field}.SHIFTL(j));
+        if (bits < j + 64 && (field >> (bits - j)) != 0) {
+          result.overflow = true;
+          break;
+        }
+      }
+    }
+    return result;
+  }
+
+  // TODO formatting
+
+  static constexpr Integer BIT_SIZE() { return {std::uint64_t{bits}}; }
+  static constexpr Integer HUGE() { return MASKR(bits - 1); }
+
+  // Returns the number of full decimal digits that can be represented.
+  static constexpr int RANGE() {
+    if (bits < 4) {
+      return 0;
+    }
+    Integer x{HUGE()}, ten{std::uint64_t{10}};
+    int digits{0};
+    while (x.CompareUnsigned(ten) != Ordering::Less) {
+      ++digits;
+      x = x.DivideUnsigned(ten).quotient;
+    }
+    return digits;
+  }
+
+  constexpr bool IsZero() const {
+    for (int j{0}; j < parts; ++j) {
+      if (part_[j] != 0) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  constexpr bool IsNegative() const {
+    return (LEPart(parts - 1) >> (topPartBits - 1)) & 1;
+  }
+
+  constexpr Ordering CompareToZeroSigned() const {
+    if (IsNegative()) {
+      return Ordering::Less;
+    } else if (IsZero()) {
+      return Ordering::Equal;
+    } else {
+      return Ordering::Greater;
+    }
+  }
+
+  // Count the number of contiguous most-significant bit positions
+  // that are clear.
```

Comment:
If this is supposed to match the semantics of Fortran, you might say so in the comment.  Same for similar fcns.

Reply:
There's a comment to that effect above the beginning of the class template definition.